[FLINK1919] add HCatOutputFormat

jamescao opened the pull request:
[FLINK1919]
Add `HCatOutputFormat` for Tuple data types for java and scala api also fix a bug for the scala api's `HCatInputFormat` for hive complex types.
Java api includes check for whether the schema of the HCatalog table and the Flink tuples match if the user provides a `TypeInformation` in the constructor. For data types other than tuples, the OutputFormat requires a preceding Map function that converts to `HCatRecords`
scala api includes check if the schema of the HCatalog table and the Scala tuples match. For data types other than scala Tuple, the `OutputFormat` requires a preceding Map function that converts to HCatRecords scala api requires user to import `org.apache.flink.api.scala`._ to allow the type be captured by the scala macro.
The Hcatalog jar in maven central is compiled using hadoop1, which is not compatible with hive jars for testing, so a cloudera hcatalog jar is pulled into the pom for testing purpose. It can be removed if not required.
java List and Map can not be cast to scala `List` and `Map`, `JavaConverters` is used to fix a bug in HcatInputFormat scala api

@chiwanpark @rmetzger 
I have changed the hcatalog jar to the apache version. That requires that I move the hcatalog module to hadoop1 profile. 
@chiwanpark 
I had made changes to most of your comment. Except for your comment regarding the verification of Exception in the tests. I feel that it's better to verify the exception at the point it's expected to be thrown. If we use method-wide annotation, we are not sure where the exception is thrown from the test method, this is not safe especially for common exception types such as IOException. I did remove the test dependency on exception error message.

chiwanpark commented:
From [this JIRA](https://issues.apache.org/jira/browse/HIVE-4460), It seems that HCatalog are deployed with Hadoop 2. Is it sure that the vanila version of HCatalog is only with Hadoop 1?

jamescao commented:
@chiwanpark 
Yes, it's sure that the apache version of HCatalog is compiled against hadoop1. Using this jar in an hadoop2 environment will lead to binary incompatibility error. I think Hive community is aware of this problem.
https://issues.apache.org/jira/browse/HIVE-7349
On the other hand. The cloudera Hcatalog jar is compiled against hadoop2, thus works with hadoop2 environment. But using cloudera jar in a hadoop1 environment will lead to binary incompatibility error. I guess one way to solve this is to add a vendor-specific hcatalog-module for users who want to use hcatalog in hadoop2 or we need to wait until hive community publish 2 artifacts in maven for both hadoop1 and hadoop2.

jamescao commented:
Any updates from the review process?

chiwanpark commented:
Oh, sorry for late. I'm reviewing this.

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class HcatInputOutputFormatITest {
```
`HCatInputOutputFormatITCase` would better for name.

chiwanpark commented on flink-staging/flink-hcatalog/src/test/scala/org/apache/flink/hcatalog/scala/test/HCatInputOutputFormatITest.scala:
```diff
@@ -0,0 +1,216 @@
+import org.junit.rules.ExpectedException
+import org.junit.{Rule, After, Before, Test}
+
+class HCatInputOutputFormatITest {
```
`HCatInputOutputFormatITCase` would be better

chiwanpark commented on flink-staging/flink-hcatalog/src/test/scala/org/apache/flink/hcatalog/scala/test/HCatInputOutputFormatITest.scala:
```diff
@@ -0,0 +1,216 @@
+
+  @Test
+  @throws(classOf[Exception])
+  def testWriteComplexTypeParition {
```
Typo: `testWriteComplexTypePartition`

chiwanpark commented on flink-staging/flink-hcatalog/src/test/scala/org/apache/flink/hcatalog/scala/test/HCatInputOutputFormatITest.scala:
```diff
@@ -0,0 +1,216 @@
+  }
+
+  def makePathASafeFileName(filePath: String): String = {
+    return new File(filePath).getPath.replaceAll("\\\\", "/")
```
Do not need `return` for Scala

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java:
```diff
@@ -0,0 +1,301 @@
+
+	private Configuration configuration;
+	//a Job is used to create OutputJobInfo
+	private Job job;
```
If there is need an explanation for a field, please use JavaDoc style (/** */).

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java:
```diff
@@ -0,0 +1,301 @@
+	private RecordWriter<WritableComparable<?>, HCatRecord> recordWriter;
+
+	//required data type by the table
+	protected TypeInformation<T> reqType;
```
Need to JavaDoc

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java:
```diff
@@ -0,0 +1,301 @@
+
+		try {
+			this.context = HadoopUtils.instantiateTaskAttemptContext(this.configuration,
+					taskAttemptID);
```
Unnecessary new line.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java:
```diff
@@ -0,0 +1,301 @@
+
+		TaskAttemptID taskAttemptID = TaskAttemptID.forName("attempt__0000_r_"
+				+ String.format("%" + (6 - Integer.toString(taskNumber + 1).length()) + "s", " ")
+				.replace(" ", "0")
```
Unnecessary new line.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java:
```diff
@@ -0,0 +1,301 @@
+
+		this.hCatOutputFormat = new org.apache.hive.hcatalog.mapreduce.HCatOutputFormat();
+		this.outputSchema = (HCatSchema) HCatUtil.deserialize(
+				this.configuration.get("mapreduce.lib.hcat.output.schema"));
```
Unnecessary new line.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java:
```diff
@@ -0,0 +1,301 @@
+		this.outputSchema = (HCatSchema) HCatUtil.deserialize(
+				this.configuration.get("mapreduce.lib.hcat.output.schema"));
+		org.apache.hive.hcatalog.mapreduce.HCatOutputFormat.
+				setSchema(this.configuration, this.outputSchema);
```
Unnecessary new line.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+	 */
+	public HCatOutputFormat(String database, String table, Map<String, String> partitionValues
+							) throws IOException
+	{
```
Unnecessary new line.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+	 */
+	public HCatOutputFormat(String database, String table, Map<String, String> partitionValues,
+							Configuration conf) throws IOException
+	{
```
Unnecessary new line.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+	 */
+	public HCatOutputFormat(String database, String table, Map<String, String> partitionValues,
+							Configuration conf, TypeInformation<T> typeInfo) throws IOException
+	{
```
Unnecessary new line.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+	{
+		super(database, table, partitionValues, conf);
+		if(!typeInfo.equals(this.reqType))
+		{
```
Unnecessary new line.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+	 * @throws IOException
+	 */
+	public HCatOutputFormat(String database, String table, Map<String, String> partitionValues,
+							Configuration conf, TypeInformation<T> typeInfo) throws IOException
```
Because of using tab character for indentation, inserting newline before starting first parameter would be better.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/scala/org/apache/flink/hcatalog/scala/HCatOutputFormat.scala:
```diff
@@ -0,0 +1,449 @@
+                                            ) extends
+HCatOutputFormatBase[T](database, table, mapAsJavaMap(partitionValues), config) {
+
+  //get the TypeInformation
```
Unnecessary comment

chiwanpark commented on flink-staging/flink-hcatalog/src/main/scala/org/apache/flink/hcatalog/scala/HCatOutputFormat.scala:
```diff
@@ -0,0 +1,449 @@
+        throw new IOException("field has different type from required")
+      }
+    }
+  }
```
`isInstanceOf` is not Scala style. For this case, pattern matching would be better. I added an exception when the type of record is not derived from Tuple class.

```scala
ti match {
    case value: CaseClassTypeInfo[_] =>
      //check the number of columns
      if (value.getArity != this.reqType.getArity) {
        throw new IOException("tuple has different arity from the table's column numbers")
      }

      for (i <- 0 until this.reqType.getArity) {
        val it = value.getTypeAt(i)
        val ot = reqType.asInstanceOf[TupleTypeInfo[_]].getTypeAt(i)

        if (!isCompatibleType(it, ot)) {
          throw new IOException("field has different type from required")
        }
      }
    case _ => throw new IOException("HCatOutputFormat only supports tuple-derived type.")
  }
```

chiwanpark commented on flink-staging/flink-hcatalog/src/main/scala/org/apache/flink/hcatalog/scala/HCatOutputFormat.scala:
```diff
@@ -0,0 +1,449 @@
+
+  //convert the type T to HCatRecord, if T is already HCatRecord, do nothing
+  override protected def TupleToHCatRecord(record: T): HCatRecord = {
+    if (ti.isInstanceOf[CaseClassTypeInfo[_]]) {
```
Need to use pattern matching.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/scala/org/apache/flink/hcatalog/scala/HCatOutputFormat.scala:
```diff
@@ -0,0 +1,449 @@
+    if (i.isInstanceOf[Map[_, _]]) i.asInstanceOf[Map[_, _]].asJava
+    else if (i.isInstanceOf[List[_]]) i.asInstanceOf[List[_]].asJava
+    else i
+  }
```
Need to use pattern matching.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/scala/org/apache/flink/hcatalog/scala/HCatOutputFormat.scala:
```diff
@@ -0,0 +1,449 @@
+      case _ =>
+        throw new IOException("Only scala tuple is allowed")
+
+    }
```
We can reimplement this with converting scala collection to java collection.

```scala
o match {
      case Tuple1(i1: AnyRef) => Seq(toJava(i1)).asJava
      case Tuple2(i1: AnyRef, i2: AnyRef) => Seq(toJava(i1), toJava(i2)).asJava
      case Tuple3(i1: AnyRef, i2: AnyRef, i3: AnyRef) =>
        Seq(toJava(i1), toJava(i2), toJava(i3)).asJava
      case Tuple4(i1: AnyRef, i2: AnyRef, i3: AnyRef, i4: AnyRef) =>
        Seq(toJava(i1), toJava(i2), toJava(i3), toJava(i4)).asJava
      ...
}
```

Returning this value and passing this collection to `DefaultHCatRecord` constructor. With this implementation, we don't need to case `record` to `AnyRef`.

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+            throw new RuntimeException("Could not create " + warehouseDir);
+        }
+
+
```
Unnecessary new lines

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+        List<Tuple3> l = d.collect();
+        Integer i = 0;
+        for(Tuple3 t: l)
+        {
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+
+        ExecutionEnvironment env = ExecutionEnvironment.createCollectionsEnvironment();
+        HCatInputFormat<Tuple3> ipf = (HCatInputFormat<Tuple3>)
+        new HCatInputFormat<Tuple3>(null, "test_table1", hiveConf).asFlinkTuples();
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+            String col1 = "a" + i;
+            String col2 = "b" + i;
+            input[i] = i + "\t" + col1 + "\t" + col2;
+        }
```
Because number of row is 3, creating input manually is more readable.

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+        ExecutionEnvironment env = ExecutionEnvironment.createCollectionsEnvironment();
+
+        DataSource<Tuple3<ArrayList<String>, HashMap<Integer, String>, ArrayList<Object>>> d =
+                env.fromCollection(l);
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+            assertEquals("2nd row", line, "a2/b2/c2\t1:v21/2:v22\td/2");
+            reader.close();
+        } catch(Exception e)
+        {
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+
+
+        List<Tuple3<ArrayList<String>, HashMap<Integer, String>, ArrayList<Object>>> l =
+                new ArrayList<>(2);
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+        ExecutionEnvironment env = ExecutionEnvironment.createCollectionsEnvironment();
+
+        DataSource<Tuple3<ArrayList<String>, HashMap<Integer, String>, ArrayList<Object>>> d =
+                env.fromCollection(l);
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+
+        //verify the result has been written to the table
+        String outputFileName = makePathASafeFileName(warehouseDir + File.separator +
+                "test_table4/c0=part0/part-m-00001");
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+            assertEquals("2nd row", line, "a2/b2/c2\t1:v21/2:v22\td/2");
+            reader.close();
+        } catch(Exception e)
+        {
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+            reader.close();
+        } catch(Exception e)
+        {
+            e.printStackTrace();
```
Why we catch exception? With this, we cannot detect test failure.

chiwanpark commented on flink-staging/flink-hcatalog/src/test/java/org/apache/flink/hcatalog/java/test/HcatInputOutputFormatITest.java:
```diff
@@ -0,0 +1,383 @@
+            @SuppressWarnings("unchecked")
+            HCatOutputFormat<Tuple3<ArrayList<String>, HashMap<Integer, String>, ArrayList<Object>>>
+                    opf1
+                    = new HCatOutputFormat<>(null, "test_table4", partitionValues, hiveConf,
```
Unnecessary new line. (For only this line, other lines are okay)

chiwanpark commented on flink-staging/flink-hcatalog/src/test/scala/org/apache/flink/hcatalog/scala/test/HCatInputOutputFormatITest.scala:
```diff
@@ -0,0 +1,216 @@
+    val partitionValues: Map[String, String] = Map("c0" -> "part0")
+
+    //wrong type, should throw exception
+    try{
```
Need a space after `try`

chiwanpark commented on flink-staging/flink-hcatalog/src/test/scala/org/apache/flink/hcatalog/scala/test/HCatInputOutputFormatITest.scala:
```diff
@@ -0,0 +1,216 @@
+      val opf =
+        new HCatOutputFormat[(List[String], Map[Int, String], List[Any], Float, Float)](null,
+          "test_table", partitionValues, hiveConf)
+    } catch{
```
Need a space after `catch`

chiwanpark commented on flink-staging/flink-hcatalog/src/test/scala/org/apache/flink/hcatalog/scala/test/HCatInputOutputFormatITest.scala:
```diff
@@ -0,0 +1,216 @@
+    try{
+      val opf =
+        new HCatOutputFormat[(List[String], Map[Int, String], List[Any], Float, Float)](null,
+          "test_table", partitionValues, hiveConf)
```
We can write this expression in two lines.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/scala/org/apache/flink/hcatalog/scala/HCatOutputFormat.scala:
```diff
@@ -0,0 +1,449 @@
+      fillList(fieldList, record.asInstanceOf[AnyRef])
+      new DefaultHCatRecord(fieldList)
+    }
+    else if (record.isInstanceOf[HCatRecord]) {
```
Unnecessary new line

chiwanpark replied:
We can implement this with matching `ti` to `WritableTypeInfo[HCatRecord]`.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/scala/org/apache/flink/hcatalog/scala/HCatOutputFormat.scala:
```diff
@@ -0,0 +1,449 @@
+    else if (record.isInstanceOf[HCatRecord]) {
+      record.asInstanceOf[HCatRecord]
+    }
+    else {
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java:
```diff
@@ -0,0 +1,301 @@
+				return new GenericTypeInfo(List.class);
+			default:
+				throw new IllegalArgumentException("Unknown data type \"" +
+						fieldSchema.getType() + "\" encountered.");
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java:
```diff
@@ -0,0 +1,301 @@
+	 * @throws java.io.IOException
+	 */
+	public HCatOutputFormatBase(String database, String table, Map<String, String> partitionValues)
+			throws IOException {
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/HCatOutputFormatBase.java:
```diff
@@ -0,0 +1,301 @@
+		this(database, table, partitionValues, new Configuration());
+	}
+
+	protected abstract HCatRecord TupleToHCatRecord(T record) throws IOException;
```
Other name such as `convertTupleToHCatRecord` would be better. The name of method should start with lower case.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	@Override
+	protected HCatRecord TupleToHCatRecord(T record) throws IOException{
```
Need a space after `IOException`

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+
+	@Override
+	protected HCatRecord TupleToHCatRecord(T record) throws IOException{
+		if(record instanceof  Tuple) {
```
Need to remove a space before `Tuple` and add a space after `if`.

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+			}
+			return new DefaultHCatRecord(fieldList);
+		}
+		else if(record instanceof HCatRecord){
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+		}
+		else if(record instanceof HCatRecord){
+			//assume the user has done the conversion in previous map step
+			return (HCatRecord)record;
```
Need a space before `record`

chiwanpark commented on flink-staging/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatOutputFormat.java:
```diff
@@ -0,0 +1,126 @@
+			//assume the user has done the conversion in previous map step
+			return (HCatRecord)record;
+		}
+		else {
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/scala/org/apache/flink/hcatalog/scala/test/HCatInputOutputFormatITest.scala:
```diff
@@ -0,0 +1,216 @@
+      line = reader.readLine
+      assertEquals("2nd row", line, "a2/b2/c2\t1:v21/2:v22\td/2\te2\t0.2")
+      reader.close
+    }
```
Unnecessary new line

chiwanpark commented on flink-staging/flink-hcatalog/src/test/scala/org/apache/flink/hcatalog/scala/test/HCatInputOutputFormatITest.scala:
```diff
@@ -0,0 +1,216 @@
+      assertEquals("1st row", line, "a1/b1/c1\t1:v11/2:v12\td/1\te1\t0.1")
+      line = reader.readLine
+      assertEquals("2nd row", line, "a2/b2/c2\t1:v21/2:v22\td/2\te2\t0.2")
+      reader.close
```
Please move `reader.close` to `finally` block.

chiwanpark commented:
Hi @jamescao, I just reviewed your PR. There are some issues to merge.

* `HcatInputOutputFormatITest.java` must use tab characters for indentation.
* There are some formatting issues which I commented.
* Some Scala codes doesn't seems Scala style.

As @twalthr said, I think also that we should split the current test cases to test cases for `HCatInputFormat` and `HCatOutputFormat`. This issues and PR should cover HCatalog output format. We can post another issue to JIRA and send PR for that.

jamescao commented:
@chiwanpark 
Thanks for all you comments! I will fix them in subsequent commits. There is also a bug fix for HcatInputformat in this pr, do you think that part should aslo be removed from this pr?

chiwanpark commented:
@jamescao Yes, I think the part (including tests) related with `HCatInputFormat` should be removed.

fhueske commented:
@jamescao good observation! The fix of #1111 should be added to the HCatOutputFormatBase as well. 
There was another fix for HadoopIOFormats (FLINK-2555, PR #1038) that is also relevant for this PR.

jamescao commented:
@chiwanpark 
pr is now updated, I pull out code related to HCatInputFormat and incorporated Flink-2555 and Flink-2617. I also change the test environment to from `CollectionsEnviroment` to  `LocalEnviroment` to cover the test of the de-serialization for the HCatOutputFormat.

fpompermaier commented:
Any update on this?
